[sled agent] Add default routes, using GZ addresses as gateways

smklein opened the pull request:
Follow-up to https://github.com/oxidecomputer/omicron/pull/1066

This ensures that L3 routing from off-device can reach the external Nexus interface.

bnaecker commented on sled-agent/src/services.rs:
```diff
@@ -267,6 +270,23 @@ impl ServiceManager {
+            // This is particularly important for accessing Nexus' external
+            // interface from off-device.
+            running_zone
+                .run_cmd(&["/usr/sbin/routeadm", "-e", "ipv6-forwarding"])
```
Does this invocation not need a `-u` flag?

smklein replied:
I figured that was unnecessary because we're doing "-u" in the next invocation.

(I tried to do it in one line - `routeadm -e ipv6-forwarding ipv6-routing -u` - and the command seemed to be sad that I used multiple options at once)


... that being said, "-e option -e option" seems to work, so I'll do that.

bnaecker reviewed (approved):
Looks good!

Just to make sure I understand, the issue here was that by hanging the IP addresses of an etherstub, they're acting like a separate L2 segment, even though they're in the same L3 subnet of the sled. Thus we need forwarding, to get the OS to deliver packets from the GZ's L2 segment to the etherstub's, and routing for the OS to automatically update routing tables to describe the interfaces used to do that forwarding. Is that a fair summary?

rmustacc commented:
So a non-global zone shouldn't need anything special here whether with maghemite or special services enabled here. What led us to this conclusion and were we setting a default gateway?

smklein commented:
> Just to make sure I understand, the issue here was that by hanging the IP addresses of an etherstub, they're acting like a separate L2 segment, even though they're in the same L3 subnet of the sled. Thus we need forwarding, to get the OS to deliver packets from the GZ's L2 segment to the etherstub's, and routing for the OS to automatically update routing tables to describe the interfaces used to do that forwarding. Is that a fair summary?

I *believe* so? The concrete issue I faced was:

- Try to ping sled agent address (in Helios GZ) from a different machine (my Linux laptop) -> Observe success
- Try to ping nexus address (in Helios non-GZ) from a different machine (my Linux laptop) -> Observe failure
- experimentally, updating the `routeadm` arguments fixed this issue.

> So a non-global zone shouldn't need anything special here whether with maghemite or special services enabled here.

Just to make sure we're on the same page - Maghemite is not currently integrated with Omicron, so this code is not assuming it's present.

> What led us to this conclusion and were we setting a default gateway?

We set routes in non-global zones here: https://github.com/oxidecomputer/omicron/blob/1593b8bdac6b5e94c1cc04beb56280103b386da2/sled-agent/src/services.rs#L232-L249

Which is implemented like this:

rmustacc commented:
> I _believe_ so? The concrete issue I faced was:
> 
>     * Try to ping sled agent address (in Helios GZ) from a different machine (my Linux laptop) -> Observe success
> 
>     * Try to ping nexus address (in Helios non-GZ) from a different machine (my Linux laptop) -> Observe failure
> 
>     * experimentally, updating the `routeadm` arguments fixed this issue.

I'd like to work with you to root cause this a bit more. In particular, the thing that's weird here is that in a normal context, we're not actually supposed to be forwarding in the non-global zone. Forwarding would suggest that we have multiple interfaces and want to take traffic that came in on one that wasn't destined for us and that it should be sent out another (or some other internal rewriting is going on with ipf which shouldn't be on the scene here). So in this context to me, the requirement to have forwarding doesn't make sense and seems like something else may be going on here.
 
> > So a non-global zone shouldn't need anything special here whether with maghemite or special services enabled here.
> 
> Just to make sure we're on the same page - Maghemite is not currently integrated with Omicron, so this code is not assuming it's present.

My point was that you said in the start of this:

> Presumably we'll have a more sophisticated solution here once maghemite is integrated, but this works in the meantime.

And that the integration of maghemite should not impact the non-global zone as it will not run there or the set of configuration that it requires.

> > What led us to this conclusion and were we setting a default gateway?
> 
> We set routes in non-global zones here:
> 
> https://github.com/oxidecomputer/omicron/blob/1593b8bdac6b5e94c1cc04beb56280103b386da2/sled-agent/src/services.rs#L232-L249
> 
> Which is implemented like this:

OK. I would have expected just a simple default route for everything to go back to the gateway. So I think we can simplify what's going on here.

smklein commented:
Robert, reading your feedback, I agree with you - routing / forwarding really *shouldn't* be necessary for the non-global zones.

I went ahead and re-did this PR - it accomplishes the same goal, but through a different mechanism. Rather than enabling this routing in non-global zones, I changed the way routes are getting set up.

Rather than routing to addresses within the AZ of the system, I implemented your suggestion, and added a **default route** which points to a corresponding **gateway address from the global zone**.

This works - I can now access Nexus from off-machine. This is kinda interesting - I basically only changed the routing tables of the non-global zones, which... _shouldn't_ have been necessary to actually *reach* those zones, but perhaps may have been necessary for ping to respond to addresses coming from outside the AZ subnet.

rcgoodfellow commented:
> Try to ping sled agent address (in Helios GZ) from a different machine (my Linux laptop) -> Observe success
> Try to ping nexus address (in Helios non-GZ) from a different machine (my Linux laptop) -> Observe failure

Do I understand correctly that in the above, your laptop was pinging from a source address outside the AZ /48, the non-global zones were set up with a route onto the AZ /48, and changing that route to be a more general default route allowed communications to succeed from your laptop?

smklein commented:
> > Try to ping sled agent address (in Helios GZ) from a different machine (my Linux laptop) -> Observe success
> > Try to ping nexus address (in Helios non-GZ) from a different machine (my Linux laptop) -> Observe failure
> 
> Do I understand correctly that in the above, your laptop was pinging from a source address outside the AZ /48, the non-global zones were set up with a route onto the AZ /48, and changing that route to be a more general default route allowed communications to succeed from your laptop?

This is accurate, as far as I understand it.

rcgoodfellow reviewed (commented):
LGTM
